### PR TITLE
First commits

### PR DESCRIPTION
**Pull request by Ojisama**
Hello,
Here are the first features asked: preview of the price, validation of the form, redirection and change calculation.
I wasn't sure about how you would like the form to handle (paid < price), so I let two ways of dealing with it: 

1. the min attribute of the "paid" input is the product price (JS) 
2. in app.py I double-check if (paid < price) and ask the user to re-enter a suitable "paid".

### CODE REVIEW

**Comment by Ojisama** [comment]
I'm sorry I can't find how to add you as an Assignee (quick Google search tells me I should have a gear icon, but it doesn't appear, cf screencap)
image

File: templates/confirmation.jinja [hunk 1]
```diff
@@ -5,16 +5,21 @@
...
+    <dd>{{ order.price }}</dd>
 
     <dt>Change Due</dt>
-    <dd>{{ change_due }}</dd>
+    <dd>{{ order.paid-order.price }}</dd>
 
     <dt>Change Denominations</dt>
-    <dd><!-- TODO --></dd>
+    {% for coin,number in change.items() %}
+    <dd>{{number}} time{% if number!=1 %}s{% endif %} {{coin}}</dd>
```
Comment (tomhamiltonstubber) [inline templates/confirmation.jinja]: This, though a display of your understanding of jinja, would have been better just as an 'x', it looks wierd with the work times

File: app.py [hunk 2]
```diff
@@ -15,24 +16,62 @@
...
+        context = {
+            'product_id': product_id,
+            'buyer': escape(request.form['buyer']),
+            'price': PRODUCTS[product_id]['price'],
+            'paid': float(request.form['paid']),
+            'date': datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        }
+
+        if context['price'] > context['paid']:
+            flash('The buyer didn\'t pay enough!', 'warning')
```
Comment (tomhamiltonstubber) [inline app.py]: Use double quotes if using a single inside string.

File: templates/listOrders.jinja [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{% extends 'base.jinja' %}
+
+{% block main %}
+  <table style="width:100%;text-align: center">
```
Comment (tomhamiltonstubber) [inline templates/listOrders.jinja]: Better to use a css file to do this much styling.

File: templates/listOrders.jinja [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      <th width="15%"> Order ID </th>
+      <th width="10%"> Product </th>
+      <th width="25%"> Buyer </th>
+      <th width="10%"> Price </th>
+      <th width="10%"> Paid </th>
+      <th width="25%"> Date <a href="/listOrders/ASC">↑</a> <a href="/listOrders/DESC">↓</a></th>
+    </tr>
+
+    {% for order in orders %}
+    <tr style="{% if loop.index0%2==0 %} background-color: #DDD; {% endif %}border-top: 1px dashed gray;">
```
Comment (tomhamiltonstubber) [inline templates/listOrders.jinja]: Check out jinja cycle:

http://jinja.pocoo.org/docs/2.9/templates/

File: templates/listOrders.jinja [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    <tr>
+      <th width="15%"> Order ID </th>
+      <th width="10%"> Product </th>
+      <th width="25%"> Buyer </th>
+      <th width="10%"> Price </th>
+      <th width="10%"> Paid </th>
+      <th width="25%"> Date <a href="/listOrders/ASC">↑</a> <a href="/listOrders/DESC">↓</a></th>
+    </tr>
+
+    {% for order in orders %}
```
Comment (tomhamiltonstubber) [inline templates/listOrders.jinja]:
```jinja
{% for order_id, order_details in orders %}
  <td style="padding:10px">{{ order_id }}</td>
  <td style="padding:10px">{{ order_details.product_id }}</td>
  ...
{% endfor %}
```

File: templates/index.jinja [hunk 6]
```diff
@@ -40,5 +41,13 @@
   </form>
   <script>
     var products = {{ products|tojson|safe }};
+
+    function refreshPrice(productIndex){
+      var price = products[productIndex].price;
+      var price_label =  document.getElementById("product-price");
```
Comment (tomhamiltonstubber) [inline templates/index.jinja]: We've imported jquery, best to use it.

In this case, you can do:
```javascript
function refreshPrice(productIndex) {
  var price = products[productIndex].price
  $('#product-price').innerHTML = price;
  $('#paid').min = price
}
```

Comment (tomhamiltonstubber) [inline templates/index.jinja]: I like what you've done with the min here btw.

File: app.py [hunk 7]
```diff
@@ -41,10 +80,38 @@ def confirmation(order_id):
...
+                           order_id=order_id, title='Order Confirmation')
+
+
+@app.route('/listOrders/<ascending>')
+def listOrders(ascending):
+    try:
+        with open(ORDER_DB) as f:
+            orders = yaml.load(f) or {}
+    except:  # the file orders.yml does not exist yet
+        orders = []
```
Comment (tomhamiltonstubber) [inline app.py]: You are declaring this as an empty set or empty list if there are no orders, pick one.

File: app.py [hunk 8]
```diff
@@ -41,10 +80,38 @@ def confirmation(order_id):
...
+
+    if len(orders) == 0:  # there are no orders available
```
Comment (tomhamiltonstubber) [inline app.py]: Just `if orders:` would be fine

File: app.py [hunk 9]
```diff
@@ -41,10 +80,38 @@ def confirmation(order_id):
...
+    return render_template('confirmation.jinja', order=order, change=change,
+                           order_id=order_id, title='Order Confirmation')
+
+
+@app.route('/listOrders/<ascending>')
+def listOrders(ascending):
+    try:
+        with open(ORDER_DB) as f:
+            orders = yaml.load(f) or {}
+    except:  # the file orders.yml does not exist yet
```
Comment (tomhamiltonstubber) [inline app.py]: Use the correct Exception here, `except IOError:`. That way you don't catch all bad errors

File: app.py [hunk 10]
```diff
@@ -41,10 +80,38 @@ def confirmation(order_id):
         orders = yaml.load(f) or {}
 
     order = orders.get(order_id)
-    if order is None:
-        pass  # TODO what do we do here?
-    # TODO other stuff has to be calculated here.
-    return render_template('confirmation.jinja', order_id=order_id, title='Order Confirmation')
+    if order is None:  # the order does not exists; the URL could've been false
+        flash('Invalid Order ID:' + order_id +
```
Comment (tomhamiltonstubber) [inline app.py]: Python string substitution here:
```python
'Invalid Order ID: %s. Please check the URL.' % order_id
```

File: app.py [hunk 11]
```diff
@@ -41,10 +80,38 @@ def confirmation(order_id):
         orders = yaml.load(f) or {}
 
     order = orders.get(order_id)
-    if order is None:
-        pass  # TODO what do we do here?
-    # TODO other stuff has to be calculated here.
-    return render_template('confirmation.jinja', order_id=order_id, title='Order Confirmation')
+    if order is None:  # the order does not exists; the URL could've been false
```
Comment (tomhamiltonstubber) [inline app.py]: If not order:

File: templates/confirmation.jinja [hunk 12]
```diff
@@ -5,16 +5,21 @@
...
-    <dd>{{ amount_paid }}</dd>
+    <dd>{{ order.paid }}</dd>
 
     <dt>Item Price</dt>
-    <dd>{{ item_price }}</dd>
+    <dd>{{ order.price }}</dd>
 
     <dt>Change Due</dt>
-    <dd>{{ change_due }}</dd>
+    <dd>{{ order.paid-order.price }}</dd>
```
Comment (tomhamiltonstubber) [inline templates/confirmation.jinja]: `order.paid-order.price` is a bad name for this, should be change_due or something relevant

**Comment by tomhamiltonstubber** [comment]
Okay I've made a load of comments, but all in all very good.

The only other general point I have is that you are displaying cash values, so display them as cash:

£1.34 and £2.40 etc.

You can correct my comments and update the PR, then I'll look again.

File: templates/listOrders.jinja (hunk 4 above)
Comment (Ojisama) [inline templates/listOrders.jinja]: Actually I wanted to use the CSS property nth-child, but I couldn't get the .css to load properly due to a browser extension. I never used the Bootstrap CSS properties, it's very useful, good to know!

**Comment by Ojisama** [comment]
I think I fixed all the mistakes / bugs from your comments!

 However, there is still one bug that I can't fix: hitting the _'Previous Page'_ button when on the order confirmation page will not display the Price automatically, even if the correct product is still selected. I tried with the `onload ` attribute in addition to the `onchange`, but it doesn't work. Any idea on how to fix this?

**Comment by tomhamiltonstubber** [comment]
Sure, look at the following:

```jinja
<option value="{{ loop.index - 1 }}" {% if loop.index - 1 == product_id %} selected {% endif %}>{{ product.name }}</option>
...

<script>
    var $product = $('#product');
    var products = {{ products|tojson|safe }};
    $product.change(function () {
      refreshPrice()
    });
    function refreshPrice() {
      var v = $product.val();
      var price = '0.00';
      if (v) {
        price = products[$product.val()].price;
      }
      $('#product-price').html('£' + price);
      $('#paid').attr({min:price});
    }
    refreshPrice();
  </script>
```

**Comment by Ojisama** [comment]
Works perfectly. I went too complicated with the `onload` attribute, a simple call to `refreshPrice();` was enough. My bad!

**Pull request closed without merging**